Formula/adguard-home.rb

```diff
@@ -0,0 +1,90 @@
+class AdguardHome < Formula
+  desc "Network-wide ads & trackers blocking DNS server"
+  homepage "https://adguard.com/adguard-home.html"
+  url "https://github.com/AdguardTeam/AdGuardHome/archive/v0.98.1.tar.gz"
+  sha256 "37779d659bb9faa1b2273f70aba4db14a77b4467888953d055547f28e8afa707"
+  head "https://github.com/AdguardTeam/AdGuardHome.git", :using => :git
+
+  depends_on "go" => :build
+  depends_on "node@8" => :build
+
+  resource "packr" do
+    url "https://github.com/gobuffalo/packr/archive/v2.0.1.tar.gz"
+    sha256 "cc0488e99faeda4cf56631666175335e1cce021746972ce84b8a3083aa88622f"
+  end
+
+  def install
+    ENV["GOPATH"] = buildpath
+    ENV["CGO_ENABLED"] = "0"
+
+    dir = buildpath/"src/github.com/AdguardTeam/AdGuardHome"
+    dir.install buildpath.children
+
+    resource("packr").stage { system "go", "install", "./packr" }
+
+    cd dir do
+      system "npm", "--prefix", "client", "install"
+      system "npm", "--prefix", "client", "run", "build-prod"
+
+      system buildpath/"bin/packr", "-z"
+      system "go", "build", "-ldflags", "-s -w -X main.version=#{version}"
+      bin.install "AdGuardHome"
+
+      (etc/"adguard-home").mkpath
+      prefix.install_metafiles
+    end
+  end
+
+  def caveats; <<~EOS
+    On the first run, you have to go through the initial configuration
+    wizard. Open your web browser and go to:
+      http://127.0.0.1:3000
+
+    After starting and configuring AdGuardHome, you will need to point your
+    local DNS server to 127.0.0.1. You can do this by going to
+    System Preferences > "Network" and clicking the "Advanced..."
+    button for your interface. You will see a "DNS" tab where you
+    can click "+" and enter 127.0.0.1 in the "DNS Servers" section.
+    By default, AdGuardHome runs on localhost (127.0.0.1), port 53,
+    balancing traffic across a set of resolvers.
+  EOS
+  end
+
+  plist_options :startup => true
+
+  def plist; <<~EOS
+    <?xml version="1.0" encoding="UTF-8"?>
+    <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+    <plist version="1.0">
+      <dict>
+        <key>Label</key>
+        <string>#{plist_name}</string>
+        <key>KeepAlive</key>
+        <true/>
+        <key>RunAtLoad</key>
+        <true/>
+        <key>ProgramArguments</key>
+        <array>
+          <string>#{bin}/AdGuardHome</string>
+          <string>-c</string>
+          <string>#{etc}/adguard-home/AdGuardHome.yaml</string>
+          <string>-w</string>
+          <string>#{etc}/adguard-home</string>
+          <string>--no-check-update</string>
+        </array>
+        <key>UserName</key>
+        <string>root</string>
+        <key>StandardErrorPath</key>
+        <string>/dev/null</string>
+        <key>StandardOutPath</key>
+        <string>/dev/null</string>
+      </dict>
+    </plist>
+  EOS
+  end
+
+  test do
+    output = shell_output("echo \"no\" | #{bin}/AdGuardHome 2>&1", 1)
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
You have quite a lot of formulas doing the *version check*. This binary is a daemon server that needs to be run with a config file and ideally as a superuser. I don't see any way of invoking a *test* that you want. There's no input/output here.

Reply:
> You have quite a lot of formulas doing the version check.

Yes, it's better then nothing. We hold newer submissions to higher standards then existing formula.

> This binary is a daemon server that needs to be run with a config file and ideally as a superuser.

Have a look at what the httpd formula is doing. That is also a server with a config file.

Reply:
@SMillerDev I've found some time to rework the test case. It now tries to verify whether the local DNS component is working.

Reply:
Thanks @jsynowiec!